pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/internal/LanguageLevelChecker.java

```diff
@@ -0,0 +1,457 @@
+                            break;
+                        }
+                    }
+                }
+            }
+            visitChildren(node, data);
+        }
+
+        @Override
+        public void visit(ASTIntersectionType node, T data) {
+            if (node.jjtGetParent() instanceof ASTCastExpression) {
+                check(node, RegularLanguageFeature.INTERSECTION_TYPES_IN_CASTS, data);
+            }
+            visitChildren(node, data);
+        }
+
+        @Override
+        public void visit(ASTCatchStatement node, T data) {
+            if (node.isMulticatchStatement()) {
+                check(node, RegularLanguageFeature.COMPOSITE_CATCH_CLAUSES, data);
+            }
+            visitChildren(node, data);
+        }
+
+        @Override
+        public void visit(ASTSwitchLabel node, T data) {
+            if (IteratorUtil.count(node.iterator()) > 1) {
+                check(node, PreviewFeature.COMPOSITE_CASE_LABEL, data);
+            }
+            visitChildren(node, data);
+        }
+
+        @Override
+        public void visit(ASTModuleDeclaration node, T data) {
+            check(node, RegularLanguageFeature.MODULE_DECLARATIONS, data);
+            visitChildren(node, data);
+        }
+
+        @Override
+        public void visit(ASTSwitchLabeledRule node, T data) {
+            check(node, PreviewFeature.SWITCH_RULES, data);
+            visitChildren(node, data);
+        }
+
+        @Override
+        public void visit(ASTVariableDeclaratorId node, T data) {
+            checkIdent(node, node.getVariableName(), data);
+            visitChildren(node, data);
+        }
+
+        @Override
+        public void visit(ASTAnyTypeDeclaration node, T data) {
+            checkIdent(node, node.getSimpleName(), data);
+            visitChildren(node, data);
+        }
+
+        private void visitChildren(JavaNode node, T data) {
+            for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+                node.jjtGetChild(i).jjtAccept(visitor, data);
+            }
+        }
+
+        private void checkIdent(JavaNode node, String simpleName, T acc) {
+            if ("var".equals(simpleName)) {
+                check(node, ReservedIdentifiers.VAR_AS_A_TYPE_NAME, acc);
+            } else if ("enum".equals(simpleName)) {
+                check(node, ReservedIdentifiers.ENUM_AS_AN_IDENTIFIER, acc);
+            } else if ("assert".equals(simpleName)) {
+                check(node, ReservedIdentifiers.ASSERT_AS_AN_IDENTIFIER, acc);
+            } else if ("_".equals(simpleName)) {
+                check(node, ReservedIdentifiers.UNDERSCORE_AS_AN_IDENTIFIER, acc);
+            }
+        }
+
+    }
+
+    private static String displayNameLower(String name) {
+        return name.replaceAll("__", "-")
+                   .replace('_', ' ')
+                   .toLowerCase(Locale.ROOT);
+    }
+
+
+    private static String versionDisplayName(int jdk) {
```

Comment:
Any reason to not use `LanguageVersion`?

Reply:
Language versions are not easy to access down this code path, because they're not known to JavaLanguageHandler